Timer: let timer go overtime

hatmajster opened the pull request:
In short: Now timer can go overtime - after ringing counter will turn red and start counting up to show when exactly did the timer ring. This way if I didn't feel alarm,or just had other stuff to do, I can just look at the watch and see "oh, so the alarm rang 2 minutes ago, fine".

Description: When timer has ended and vibrated, the counter turns red and start counting up. It will turn off if user will press "stop" or close Timer application. It will not turn off if the screen will just go blank (but I've found during testing that notifications can turn red timer off - I don't know how to gracefully go around it, but it should be doable. For now left it as is). The timer will also turn off after 99:59 overtime, just to not mess with display.

This is actually most common behavior in modern timers I think. For example all smartphones have this I think.

This PR will conflict with https://github.com/JF002/InfiniTime/pull/554 and https://github.com/JF002/InfiniTime/pull/407 as clang-format reformatted all files, and also I did myself some minor cleanings.

PS1: I changed `GetTimeRemaining` to give seconds instead of milliseconds as that's what was used anyway, and changed the function name to `GetSecondsRemaining`. Hopefully with the name changed it won't break anyone's custom code.

PS2: currently counter in Timer shows `floor(remainingSeconds)`, wouldn't it be better if it would show `ceil(remainingSeconds)`? I think that's what @Riksu9000 also did in https://github.com/JF002/InfiniTime/pull/554

Riksu9000 commented:
>PS2: currently counter in Timer shows floor(remainingSeconds), wouldn't it be better if it would show ceil(remainingSeconds)? I think that's what @Riksu9000 also did in #554

Correct.

Ringing will be added in #342/#544, which can be used here to make timer ending more obvious and less likely to get missed. Still I do think that showing overtime is a good idea.

Riksu9000 reviewed (commented):
If only there was a bit more space to add a minus. Maybe instead it could read overtime below the time, and maybe blink slowly.

It should be very easy to add the ringing now with the overtime state and stop button, if you'd like to take a look at that.

Maybe even style the stop button like in the Alarm app, so it would be consistent.

hatmajster commented:
@Riksu9000 I think I've done all You asked, but I am not sure about time blinking, so I left it for now. In summary:
 - implemented `stop()` function
 - added ringing
 - matched app layout to Alarm

@JF002 I think this is one of the simplest most thumbed up change;) Can You review/give Your opinion (or both);) I've even added kinda crappy movie now, so hopefully its visible what it does ^^

hatmajster commented on src/displayapp/screens/Timer.cpp:
```diff
@@ -94,16 +118,24 @@ void Timer::OnButtonEvent(lv_obj_t* obj, lv_event_t event) {
+          minutesToSet = 0;
+          secondsToSet = 0;
+          lv_obj_set_style_local_text_color(time, LV_LABEL_PART_MAIN, LV_STATE_DEFAULT, LV_COLOR_GRAY);
+          lv_label_set_text_static(time, "00:00");
```
Hopefully now its better

Riksu9000 commented:
The layouts nice, but in the alarm app the stop button turns red when it's ringing. This is what I meant.

![IMG_20211105_104210](https://user-images.githubusercontent.com/37774658/140483792-a92158a9-a625-4b0d-ae57-ff312f8f5517.jpg)

This is what I had in mind about the overtime, and have just the text blink. But that's just an idea. I haven't tested it if it would be any good.

![overtime](https://user-images.githubusercontent.com/37774658/140483915-3c302039-be85-4405-a2b6-f298dd5cc170.png)

Itai-Nelken commented:
Works fine and is a great improvement.

Riksu9000 commented:
I think the ringing might actually make the overtime redundant. It's unlikely that the user would miss the timer ending now since the watch is worn on the wrist, compared to a timer on a phone which could get misplaced. So maybe we only need the ringing after all.

M4xTheMan commented:
I'd be interested in the overtime thing nevertheless. 

On January 1, 2022 4:30:59 PM GMT+01:00, Riku Isokoski ***@***.***> wrote:
>I think the ringing might actually make the overtime redundant. It's unlikely that the user would miss the timer ending now since the watch is worn on the wrist, compared to a timer on a phone which could get misplaced. So maybe we only need the ringing after all.
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/InfiniTimeOrg/InfiniTime/pull/557#issuecomment-1003574313
>You are receiving this because you are subscribed to this thread.
>
>Message ID: ***@***.***>

trman commented:
> I think the ringing might actually make the overtime redundant. It's unlikely that the user would miss the timer ending now since the watch is worn on the wrist, compared to a timer on a phone which could get misplaced. So maybe we only need the ringing after all.

i thinks the overtime is usefull because it allow to see how much have overtime (if you cook something , overtime of 1 minute is not thze same than 15 or more)

Riksu9000 commented:
I'm thinking that because the timer ending is now very obvious, the user wouldn't miss it in the first place. Because of this, the timer won't be left going into overtime accidentally.

As for letting it go to overtime intentionally, I don't think people will want to keep the watch vibrating on their wrist for minutes just to keep track of overtime. Though if this is important, maybe there should be a timeout and/or a mute button, but let's leave that for another PR.

Itai-Nelken commented:
Maybe simply stop the vibrations after a certain amount of time, for example 10 seconds.

hatmajster commented:
To me overtime is an important feature trman said, during cooking, baking I take it into consideration constantly. Its maybe even more important than missing the timer as I usually just check it constantly. I added now the 10 seconds for stopping vibrations, as indeed ringing all the time was a bit of annoyance.
I am still wondering if overtime timer should stop after closing application. It is sometimes happening and then the information lost... On the other hand I don't like clicking on watch that much, and everything currently is just so clean;)

Riksu9000 commented:
Okay, I'm willing to approve this PR with or without the 10 second timeout, as it's an improvement either way. I may make further improvents to this in the future.

If you want to fix the timeout, you'll need to disable sleeping while the timer is ringing. This is because the apps don't run while the screen is off, so `Refresh()` isn't called. Alternatively I think you can use FreeRTOS timers, which keep running while the screen is off. You can take a look at #945, where I made something similar for the Alarm app. Now I'm wondering which solution is better..

hatmajster commented:
To me it doesn't matter much if the screen should be on or off during vibration, I can have/do it either way. Right now in newest commit, I implemented vibration for 10 times in motorController. Seems to be working good, and I like this way. You, Riksu could reuse it in Alarm or any other app the same way.
However, it introduces a little rare race, so it maybe better to do it differently (like You already did), do outside of this PR for now or scrap completely. The problem is when the timer is in overtime and a call will happen, it will often stop ringing. This is because Timer app and Notifications app stops vibration on closing. If timer/notification don't sync we have stopped ringing and can miss a call:/
Now that I think of it I don't know if previous implementation was secured against this anyway... I think not.

Riksu9000 commented:
Hmm, I think I would rather have the app control the ringing directly rather than leave it being controlled separately in MotorController without feedback.

There is already a bug where the ringing stops when receiving two call notifications for example, so I don't think it was caused by your changes.

mashuptwice commented:
Note: This PR is open so long that in the meantime the alarm apps design got matched to the timer. @hatmajster do you plan to finish the work on this?

Riksu9000 commented:
I would like to see some changes made, but there are also a lot of conflicts making it difficult to update and I'd have preferred the overtime feature to be separate.
